filestore/filepath.go

```diff
@@ -0,0 +1,520 @@
+}
+
+func (azure *AzureFilepath) ToURI() string {
+	return fmt.Sprintf("%s%s@%s.dfs.core.windows.net/%s", azure.scheme, azure.bucket, azure.StorageAccount, azure.key)
+}
+
+// **NOTE**: Due to Azure Blob Storage's unique URI format, we need to re-implement this method
+// on the derived type to ensure we can properly handle the `bucket` field.
+func (azure *AzureFilepath) ParseFilePath(fullPath string) error {
+	u, err := url.Parse(fullPath)
+	if err != nil {
+		return fmt.Errorf("could not parse full path '%s': %v", fullPath, err)
+	}
+	// Our scheme is the protocol + "://", so we need to suffix the scheme with "://"
+	// to ensure the comparison works.
+	scheme := fmt.Sprintf("%s://", u.Scheme)
+	err = azure.FilePath.checkSchemes(scheme)
+	if err != nil {
+		return err
+	}
+	azure.FilePath.scheme = scheme
+	azure.FilePath.bucket = u.User.String()              // The container will be in the User field due to the format <scheme>://<container>@<storage_account>
+	azure.StorageAccount = strings.Split(u.Host, ".")[0] // The host will be in the format <storage_account>.dfs.core.windows.net
+	azure.FilePath.key = strings.TrimPrefix(u.Path, "/")
+	if azure.FilePath.Ext() == "" {
+		azure.FilePath.isDir = true
+	} else {
+		azure.FilePath.isDir = false
+	}
+	return nil
+}
+
+func (azure *AzureFilepath) ParseDirPath(fullPath string) error {
+	err := azure.ParseFilePath(fullPath)
+	if err != nil {
+		return err
+	}
+	azure.FilePath.isDir = true
+	lastElem := filepath.Base(azure.FilePath.key)
+	if filepath.Ext(lastElem) != "" {
+		azure.FilePath.key = filepath.Dir(azure.FilePath.key)
+	}
+	return nil
+}
+
+func (azure *AzureFilepath) Validate() error {
+	if azure.scheme != "abfss://" {
+		return fmt.Errorf("invalid scheme '%s', must be 'abfss://'", azure.scheme)
+	}
+	if azure.StorageAccount == "" {
+		return fmt.Errorf("storage account cannot be empty")
+	}
+	if azure.bucket == "" {
+		return fmt.Errorf("bucket cannot be empty")
+	} else {
+		azure.bucket = strings.Trim(azure.bucket, "/")
+	}
+	if azure.key == "" {
+		return fmt.Errorf("key cannot be empty")
+	} else {
+		azure.key = strings.Trim(azure.key, "/")
+	}
+	azure.isValid = true
+	return nil
+}
+
+type GCSFilepath struct {
+	FilePath
+}
+
+func (gcs *GCSFilepath) ToURI() string {
+	return fmt.Sprintf("%s%s/%s", gcs.scheme, gcs.bucket, gcs.key)
+}
+
+func (gcs *GCSFilepath) Validate() error {
+	if gcs.scheme != "gs://" {
+		return fmt.Errorf("invalid scheme '%s', must be 'gs://'", gcs.scheme)
+	}
+	if gcs.bucket == "" {
+		return fmt.Errorf("bucket cannot be empty")
+	} else {
+		gcs.bucket = strings.Trim(gcs.bucket, "/")
+	}
+	if gcs.key == "" {
+		return fmt.Errorf("key cannot be empty")
+	} else {
+		gcs.key = strings.Trim(gcs.key, "/")
+	}
+	gcs.isValid = true
+	return nil
+}
+
+type HDFSFilepath struct {
+	FilePath
+}
+
+func (hdfs *HDFSFilepath) Validate() error {
+	return fmt.Errorf("not implemented")
+}
+
+type LocalFilepath struct {
+	FilePath
+}
+
+type FilePathGroupingType string
```

Comment:
Can we comment what this is?

Reply:
Don't love this abstraction, but need to understand what this does to have a better proposal